### ISSUE TITLE
Enum naming strategy should allow different naming conventions used in code

### ISSUE THREAD

**Issue by hajdamak** [FasterXML/jackson-databind#4675]
### Is your feature request related to a problem? Please describe.

Currently `CamelCaseStrategy` in `EnumNamingStrategies` assumes that enum entires in source code are named using UPPER_SNAKE_CASE convention, for example: `enum Size { VERY_BIG }`.
However in Kotlin quite often UpperCamelCase is used, for example `enum class Size { VeryBig }`.   `CamelCaseStrategy` in `EnumNamingStrategies` does not work with such naming.

### Describe the solution you'd like

Strategies in `EnumNamingStrategies` should work with enum entries written in code using different naming conventions like UpperCamelCase .

**Comment by cowtowncoder** [comment]
I am not 100% sure if this can be done reliably, but if so, sounds like a reasonable idea.

**Comment by JooHyukKim** [comment]
I believe naming conventions wrt Enum is somewhat controversial.
Here is reference to [discussion in Kotlin](https://discuss.kotlinlang.org/t/enum-constants-uppercase-or-camecase/13984).

As far as I know it is conventional to use
- UPPER_SNAKE_CASE in **Java**
- 50/50 debate on UPPER_SNAKE_CASE vs PascalCase in **Kotlin**

So it might be safer off to provide extension points instead of declaring what Enum naming would be.
But as @cowtowncoder mentioned, idk how it could be reliable, technically, since now we would have naming strategies two-ways, in-and-out.

**Comment by yihtserns** [comment]
Sounds like something that https://github.com/FasterXML/jackson-module-kotlin can/should(?) support out-of-the-box:
```
internal class KotlinAnnotationIntrospector(...) {
   ...
   override fun findEnumNamingStrategy(config: MapperConfig<*>?, ac: AnnotatedClass?): Any {
      return PascalCaseOrCamelCaseStrategy()
   }
}
```

**Comment by cowtowncoder** [comment]
@yihtserns Perhaps, although if so, should only apply to Kotlin Enum types? And be configurable to allow disabling such logic.

**Comment by yihtserns** [comment]
> should only apply to Kotlin Enum types?

I think so far nobody from Java side has ever requested such thing, right?  The reason this issue was created is because only Kotlin side has such "double standard". 🤐

**Comment by cowtowncoder** [comment]
Right, if there are reliable means to detect that information. Similar to how Kotlin special sauce avoided (I think?) for POJOs.
